### PR TITLE
More Structure in Error Object

### PR DESCRIPTION
**Pull request by FrankEssenberger**
Basis for discussion. I will schedule a meeting.

### CODE REVIEW

File: knowledge-base/adr/structure-of-error-object.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+```typescript
+const [buPa, req, res]: [BusinessPartner, Request, Response] = await BusinessPartner.requestBuilder().getAll().executeRaw(destination);
+```
+
+One could include a `httpNoThrow` flag to the `execute` methods. 
+If switched on the request builder will not throw HTTP related error but include them in the return. 
+
+```typescript
+const [buPa, httpError]: [BusinessPartner, Request, Response] = await BusinessPartner.requestBuilder().httpNoThrow().getAll().execute(destination);
```
Comment (jjtang1985) [inline knowledge-base/adr/structure-of-error-object.md:122]: Maybe adjust the type? `const [buPa, httpError]: [BusinessPartner, Request, Response]`

File: knowledge-base/adr/structure-of-error-object.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
+# More Structure in Error Object
+
+## Status
+
+proposed
```
Comment (marikaner) [inline knowledge-base/adr/structure-of-error-object.md:5]: decided

File: knowledge-base/adr/structure-of-error-object.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    name: string;
+    statucCode:number
+    message: string;    
+}
+```
+This ADR decides the way to improve the situation for the customer.
+
+## Options
+
+### Option A: Extending the Error Interface/Object
```
Comment (marikaner) [inline knowledge-base/adr/structure-of-error-object.md:62]:
```suggestion
### Option A: Extending the Error interface/object
```

File: knowledge-base/adr/structure-of-error-object.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+- Typed information specific to error: Http, IO, ...
+- Detailed information like HttpStatus etc can be accessed easily.
+
+Cons:
+- Not very JavaScripty
+- We need to find the right type in the `errorWithCause` -> big exception mapper factory
+- The consumer needs some kind of "down cast" to find the actual error type returned (catch blocks, instanceof check)
+This becomes even more problematic since there are no `throws declaration` possible.
+
+### Option B: Include the Root cause error Message
```
Comment (marikaner) [inline knowledge-base/adr/structure-of-error-object.md:87]:
```suggestion
### Option B: Include the original error
```

File: knowledge-base/adr/structure-of-error-object.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+We could at least include  a `root` property to the error:
+
+```typescript
+class ErrorWithRoot extends Error {
+    name: string;
+    message: string;
+    stack?: string;
+    originalError?:Error   
+    get rootError():Error|undefined{}
+    //recursibly check if originalError is of type ErrorWithRoot and return the first non  ErrorWithRoot error in this chain.
```
Comment (marikaner) [inline knowledge-base/adr/structure-of-error-object.md:98]:
```suggestion
    originalError?: Error;
    get rootError(): Error | undefined {
      //recursibly check if originalError is of type ErrorWithRoot and return the first non  ErrorWithRoot error in this chain.
    }
```

File: knowledge-base/adr/structure-of-error-object.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+}
+```
+
+the variable `rootMessage` would contain `Request failed with status code 400` in the example above.
```
Comment (marikaner) [inline knowledge-base/adr/structure-of-error-object.md:102]: Please update this sentence ;)

File: knowledge-base/adr/structure-of-error-object.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+Pros:
+- Information on Http Errors typed
+- No casting needed 
+
+Cons:
+- Errors are handled differently (Http vs. deserializing errors for example).
+
+### Option D: Introduce `failable` return types
+
+A pattern I found [here](https://medium.com/@dhruvrajvanshi/making-exceptions-type-safe-in-typescript-c4d200ee78e9) is a failable type:
```
Comment (marikaner) [inline knowledge-base/adr/structure-of-error-object.md:138]: I would avoid "I" and personal opinions in an ADR, I think that is irrelevant.
```suggestion
This pattern](https://medium.com/@dhruvrajvanshi/making-exceptions-type-safe-in-typescript-c4d200ee78e9) is a failable type:
```

File: knowledge-base/adr/structure-of-error-object.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+}
+``` 
+
+Introducing such a types is a bit like a throw declaration and you can list the exception which are appearing.
+
+```typescript
+function execute():Failable<BusinessPartner,AxiosError| ParseError | Error>
+```
+
+Nothe that his also goes a bit in the direction of checked exceptions but you have union types and can define type aliases which avoid that you have to change a zillion files you add one more error cause.
```
Comment (marikaner) [inline knowledge-base/adr/structure-of-error-object.md:155]:
```suggestion
Note that his also goes a bit in the direction of checked exceptions but you have union types and can define type aliases which avoid that you have to change a zillion files you add one more error cause.
```

File: knowledge-base/adr/structure-of-error-object.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+We take option B.
+Check convention on MDN how what to throw i.e. should you extend from the error object. 
+Also adjust the name `ErrorWithRoot` to something better.
+
+The colleagues from exchange rate lib had problems with the [prototype chain](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-newtarget).
+Just consider these kind of problems.
+
+## Consequences
+
+none
```
Comment (marikaner) [inline knowledge-base/adr/structure-of-error-object.md:178]: I guess this needs an update ;)

**Review by marikaner — APPROVED** [review]
Just minor stuff, maybe you can read through it again and check that the information is up to date and then merge. I already approved, so you can do this quickly. It would be great if you could also post this to the issue (I can also do it if you prefer).

**Pull request merged**